### ISSUE TITLE
Overriding fields to be final

### ISSUE THREAD

**Issue by Levi-Lesches** [dart-lang/language#1390]
Consider this code: 
```dart
class A {
  bool _temp;
  bool get temp => _temp;
  set temp(bool value) {
    print("Changing A.temp to $value");
    _temp = value;
  }
  
  A(this._temp);
}

class B extends A {
  @override
  final bool temp; 
  
  B(this.temp) : super(temp);
}

void main() {
  final A a = A(true);
  a.temp = false;  // "Changing A.temp to false"
  print(a.temp);  // false
  
  final B b = B(true);
  b.temp = false;  // "Changing A.temp to false"
  print(b.temp);  // true
}
```

Based on the print statements, it seems that Dart is referring to A's setter because B doesn't have one. I can confirm this because when I changed B.temp from final to a getter and setter like A has, B's print statements ran instead. But this behavior is unintuitive, and I would expect this to either error or B's getter to return whatever A's setter received.

**Comment by leafpetersen** [comment]
The behavior you're describing is, for better or worse, working as intended.  A pair of pretty fundamental design decisions made early on with Dart is that:
  - Setters and getters are almost entirely uncoupled
  - Fields and setters/getters are interchangeable

That said, this seems like a really good candidate for a lint/hint.  It feels to me that any time you are:
  - overriding a setter/getter pair with a final field
  - overriding a mutable field with a final field

you are highly likely to not be doing so intentionally.  It's a little less clear if you are overriding only one of a setter/getter pair, but that also seems a little suspect.

cc @pq @bwilkerson @munificent @natebosch @srawlins for thoughts on a hint/lint here?

**Comment by Levi-Lesches** [comment]
I believe there actually is a lint for this, although it's very generalized: [overridden_fields](https://dart-lang.github.io/linter/lints/overridden_fields.html).

The intended behavior _is_ understandable (which is how I knew where to put the print statements), but I feel since this is SO unintuitive and may not be discovered until there's a bug down the line, maybe this should be a compile-time error? Especially since including `final`, in either the superclass or the subclass will result in undesired behavior. Maybe whenever you override a field with a getter/setter, or override a getter/setter with a field it should be an error?

**Comment by lrhn** [comment]
There can be good reasons for overriding a field, even a mutable field, with a getter (for example to log access, in which case it typically will read `super.field` afterwards, so the field isn't lost). That even makes sense when you only override the one of the getter and setter.

There can be reasons to override getters/setters with a field (perhaps the getter/setter is a default implementation which always returns `null` and you're overriding it with a real field).

In general, it should never matter whether something is a field or a getter/setter. Making it an error to override a field with getter, but not a getter with a getter, will make it a breaking change for the superclass to change a getter to a field. That should not be a breaking change.

All in all, I think we might be able to heuristically detect some undesired cases, but it won't be as simple as disallowing some getter/setter/field override combinations unconditionally. It would have to look at the actual content and context of the overrides, and it would most likely be at most a hint.

**Comment by eernstg** [comment]
I think we can say that Dart is optimized for extensibility and not for constraints when it comes to 'properties'.

This choice is supported by the decision to introduce getters and setters implicitly for instance variables (if final: only a getter), and this means that we do _not_ have the rule of thumb (known from, e.g., Java) that every instance variable `T x` must be private and there must be a `T getX()` and/or a `void setX(T)` if it is intended to be accessed from any "non-trivial distance". The `getX`/`setX` declarations could be public, package-private, protected, etc., but the point is that we are protecting the implementation of the property `x` such that we can change our minds about storing a value in memory or computing the value on demand. Dart ensures that we abstract away from this implementation choice. This also means that Dart can use a concise syntax (`o.x` and `o.x = ...`) for computed properties.

Constraints weren't prioritized as much. In particular, getters and setters are used uniformly to read and write instance variables, and they may be overridden freely with implicitly induced as well as explicitly written getters/setters. So even the useless or error-prone combinations can be used, e.g., combinations where memory is allocated but never used, because an instance variable is overridden and no super-invocations exist.

However, constraints to avoid special cases that are useless or error-prone fit very well as a lint or a hint. They are not in the language specification, and lints can be enabled or disabled, so there is a considerable amount of freedom in the selection and treatment of such undesirable cases.

I would consider hinting the situation where storage is allocated and never used:

```dart
class A1 { int i; }
class A2 extends A1 { int i; } // No invocations of `super.i` getter or setter.
class A3 extends A1 {
  // Again no super-invocations, can't use the storage of `A1.i`.
  int get i => 42;
  set i(int value) {}
}
class A4 extends A1 {
  int get i => 42;
  // We can store new values into `A1.i`, but we can't read them.
}
```

I guess this amounts to overriding an implicitly induced getter, and never calling it via a super-invocation (in any instance member of that class). This implies that we can't _use_ that storage, even though we may actually be able to store something in there.

Similarly, we should probably hint if we can read a particular storage location, but we can't write to it. So it will have to have an initial value from an initializer or from each generative constructor, or it needs to have a nullable type. This case will cover the situation "override a getter/setter (written or implicit) by a final instance variable" mentioned [here](https://github.com/dart-lang/language/issues/1390#issuecomment-754345144), except that it won't complain if both the overridden getter and the overridden setter are invoked in super-invocations somewhere.

On the other hand, I don't think it's useful to hint just because a particular implementation is overridden and made uncallable:

```dart
class B1 { int get i => /*...something, presumably useful...*/; }
class B2 extends B1 { int i; } // OK, even without a super-invocation.
class B3 extends B1 { int get i => 43; } // Ditto.
```

**Comment by leafpetersen** [comment]
> All in all, I think we might be able to heuristically detect some undesired cases, but it won't be as simple as disallowing some getter/setter/field override combinations unconditionally. It would have to look at the actual content and context of the overrides, and it would most likely be at most a hint.

While I agree that there are specific instances that are useful, I also I identified a specific list of combinations that I thought were problematic in my comment above.

**Comment by leafpetersen** [comment]
To expand a bit on my previous comment: I think the responses are fixating on the "setter/getter" vs "field" aspect, whereas what I think is most relevant is having a thing which affords both setting and getting overridden by a final field.

**Comment by natebosch** [comment]
I agree that the 2 specific cases you described are likely to have few or no false positives. I can't think of a good reason to override a writable property with a final field, but it's easy to do by mistake. If someone has a valid reason to use this pattern and doesn't want to see the hint I think we could allow suppressing it by making the behavior explicit. I'd probably argue for in a code review anyway, even if only as a place to hang a comment.

```dart
  @override
  final bool temp;
  @override
  // A hopefully convincing argument for using this pattern...
  set temp(bool value) => super.temp = value;
```

To double check we could search for existing occurrences of this pattern and see if any look intentional.

**Comment by Levi-Lesches** [comment]
> I believe there actually is a lint for this, although it's very generalized: [`overridden_fields`](https://dart-lang.github.io/linter/lints/overridden_fields.html). However Dartpad does not show the lint.

Personally, I always took this lint with a grain of salt because of the explanation: 

> Overriding fields is almost always done unintentionally. Regardless, it is a bad practice to do so.

I think it might be appropriate to use this lint, and just add more details as to **why** it's "bad practice". That is, unless that lint is referring to general OOP principles and not this behavior.

**Comment by eernstg** [comment]
I think [`overridden_fields`](https://dart-lang.github.io/linter/lints/overridden_fields.html) rules out some perfectly meaningful designs, e.g., subclasses where a setter invocation should be intercepted, perhaps because setting the variable should be accompanied by some adjustments to other parts of the state, it should be recorded (logged, added to an undo stack, ...), it should be verified that the new value is appropriate, etc. Similarly, a subclass might have reasons to intercept getter invocations. I think it is unnecessarily limiting to decide up front that these designs are prohibited.

```dart
class  A {
  int i;
}

class B extends A {
  set i(value) {
    ... /* interception code: validate, adjust state, ... */
    super.i = value;
  }
}
```

This is exactly the reason why I [suggested](https://github.com/dart-lang/language/issues/1390#issuecomment-755005244) that a hint against undesirable overrides should focus on the underlying problem: A storage area is allocated in the object, but it is never used (because we can't write anything into it, or we can write but we can't read anything out of it). The above examples _do_ allow us to read and write the storage, so they won't trigger that hint.

**Comment by lrhn** [comment]
The problematic cases Leaf mentioned above are when overriding a mutable property (getter and setter, maybe only if non-abstract) with a final field.
I agree that that does seem awkward. Arguably, it can be just as awkward to override such a property with just a getter, if that getter is oblivious to the field being mutable. It's just harder to detect.

Disallowing such an override would mean that adding a setter to a class can break subclasses. That's potentially a good thing, since making a property settable is an important change that subclasses probably need to know about. If they just keep going with a a getter which ignores the setter, things might break.

That's true for any API change which can affect the object state. Adding a *method* which can mutate state which was previously immutable can also break subclasses relying on the state not changing. 

A final field overriding a mutable property is a good heuristic for detecting an oblivious override. It's *probably* a mistake. Lots of other things are also probably mistakes, but not as easily recognizable (although anything shadowing a field getter so that it's no longer accessible in any way, so not referenced by any `super.getter` invocations, is probably also a red flag.)

tl;dr: What Erik said. The underlying symptom is state being completely inaccessible. A final field overriding a mutable property is a good place to start looking for that problem.

**Comment by munificent** [comment]
> That said, this seems like a really good candidate for a lint/hint. It feels to me that any time you are:
> 
> * overriding a setter/getter pair with a final field
> * overriding a mutable field with a final field
> 
> you are highly likely to not be doing so intentionally. It's a little less clear if you are overriding only one of a setter/getter pair, but that also seems a little suspect.

+1 for linting this. I think the overridden_fields lint is a little too broad but these two cases seem like they are almost always erroneous.

**Comment by Levi-Lesches** [comment]
> I think overridden_fields rules out some perfectly meaningful designs, e.g., subclasses where a setter invocation should be intercepted...

> I think the overridden_fields lint is a little too broad

Completely agree with the rest of the points, but just to clarify, the `overriding_fields` lint does not warn against overriding getters or setters, just overriding the field itself. 

This will not present warnings:
```dart
class A {
	int field = 0;
}

class B extends A {
	@override
	int get field => 1;
}
```
But this will: 
```dart
class A {
	int field = 0;
}

class B extends A {
	@override
	int field = 1;
}
```

**Comment by lrhn** [comment]
This seems to be converging towards lints or warnings, so should it be moved to the SDK repository?
